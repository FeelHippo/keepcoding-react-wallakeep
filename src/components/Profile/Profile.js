import React from 'react';

import Layout from '../Layout/Layout';
import FormControl from '@material-ui/core/FormControl';
import Button from '@material-ui/core/Button';
import MaterialUiInput from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import SaveIcon from '@material-ui/icons/Save';
import DeleteIcon from '@material-ui/icons/Delete';

import Form, { Input } from '../Form';

import Session from '../../models/Session';
import imagePhoto from '../../assets/images/user.png';

import './Profile.css';

export default function Profile({
  session,
  saveSession,
  enqueueSnackbar,
  userLogout,
}) {
  const handleSubmit = value => {
    const { name, surname, maxAdverts } = value;
    const parsedMaxAdverts = parseInt(maxAdverts);
    // Genero sesión y la guardo en LS
    saveSession(
      new Session(name, surname, session.apiUrl, parsedMaxAdverts),
      true,
    );
    enqueueSnackbar('Perfil de usuario actualizado correctamente.', {
      variant: 'success',
    });
  };

  return (
    <Layout sectionTitle="Perfil de usuario">
      <Form
        initialValue={session}
        validate={({ name, surname, maxAdverts }) => {
          const parsedMaxAdverts = parseInt(maxAdverts);
          if (
            !name ||
            !surname ||
            !Number.isInteger(parsedMaxAdverts) ||
            parsedMaxAdverts <= 0
          ) {
            return 'Rellene todos los campos del formulario';
          }
        }}
        onSubmit={handleSubmit}
        onError={error => enqueueSnackbar(error, { variant: 'error' })}
        autoComplete="off"
        className="Profile__Form"
      >
        <div className="Profile_Picture">
          <img src={imagePhoto} alt="user_avatar" />
        </div>
        <FormControl fullWidth className="Profile__FormControl">
          <InputLabel shrink htmlFor="type">
            Nombre
          </InputLabel>
          <Input name="name" type="text" component={MaterialUiInput} />
        </FormControl>
        <FormControl fullWidth className="Profile__FormControl">
          <InputLabel shrink htmlFor="type">
            Apellido
          </InputLabel>
          <Input name="surname" type="text" component={MaterialUiInput} />
        </FormControl>
        <FormControl fullWidth className="Profile__FormControl">
          <InputLabel htmlFor="maxAdverts">
            Anuncios por página (Home)
          </InputLabel>
          <Input
            name="maxAdverts"
            type="number"
            min={1}
            max={20}
            component={MaterialUiInput}
          />
        </FormControl>
        <div className="Profile__Footer">
          <Button
            type="submit"
            variant="contained"
            color="primary"
            startIcon={<SaveIcon />}
            className="ButtonWallakeep ButtonWallakeep__Green"
          >
            Guardar
          </Button>
          <Button
            type="button"
            variant="contained"
            color="secondary"
            onClick={userLogout}
            startIcon={<DeleteIcon />}
          >
            Borrar
          </Button>
        </div>
      </Form>
    </Layout>
  );
}
