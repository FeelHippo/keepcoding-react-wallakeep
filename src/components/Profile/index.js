import { withSnackbar } from 'notistack';

import Profile from './Profile';
import { compose } from '../../utils/Compose';
import withSession from '../../hocs/withSession';

export default compose(withSession, withSnackbar)(Profile);
