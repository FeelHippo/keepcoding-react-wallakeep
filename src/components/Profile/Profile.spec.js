import React from 'react';
import { shallow } from 'enzyme';

import Profile from './Profile';

describe('Profile', () => {
  const props = {
    session: {
      name: 'name',
      surname: 'surname',
      maxAdverts: 4,
      apiUrl: 'apiUrl',
    },
    saveSession: jest.fn(),
    enqueueSnackbar: jest.fn(),
    userLogout: jest.fn(),
  };

  const wrapper = shallow(<Profile {...props} />);

  it('should render a Profile component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('saving profile', () => {
    const preventDefault = jest.fn();
    wrapper.find('form').simulate('submit', { preventDefault });

    it('should prevent default form submission', () => {
      expect(preventDefault).toHaveBeenCalled();
    });

    it('should save session', () => {
      expect(props.saveSession).toHaveBeenCalledWith(
        expect.objectContaining(props.session),
        true,
      );
    });

    it('should launch a succesfull snackBar', () => {
      expect(props.enqueueSnackbar).toHaveBeenCalledWith(
        expect.stringContaining('actualizado correctamente'),
        expect.objectContaining({ variant: 'success' }),
      );
    });

    it('should launch an error snackBar if name is empty', () => {
      props.enqueueSnackbar.mockClear();
      wrapper.setState({ name: '' });
      wrapper.find('form').simulate('submit', { preventDefault });
      expect(props.enqueueSnackbar).toHaveBeenCalledWith(
        expect.stringContaining('Rellene'),
        expect.objectContaining({ variant: 'error' }),
      );
    });
  });

  describe('deleting profile', () => {
    wrapper
      .find('.Profile__Footer')
      .childAt(1)
      .simulate('click');

    it('should logout user', () => {
      expect(props.userLogout).toHaveBeenCalled();
    });
  });
});
