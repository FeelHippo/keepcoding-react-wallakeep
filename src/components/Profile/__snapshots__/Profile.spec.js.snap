// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should render a Profile component 1`] = `
<Layout
  sectionTitle="Perfil de usuario"
>
  <form
    autoComplete="off"
    className="Profile__Form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="Profile_Picture"
    >
      <img
        alt="user_avatar"
        src="user.png"
      />
    </div>
    <WithStyles(ForwardRef(FormControl))
      className="Profile__FormControl"
      fullWidth={true}
    >
      <WithStyles(ForwardRef(InputLabel))
        htmlFor="type"
        shrink={true}
      >
        Nombre
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        name="name"
        onChange={[Function]}
        required={true}
        type="text"
        value="name"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      className="Profile__FormControl"
      fullWidth={true}
    >
      <WithStyles(ForwardRef(InputLabel))
        htmlFor="type"
        shrink={true}
      >
        Apellido
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        name="surname"
        onChange={[Function]}
        required={true}
        type="text"
        value="surname"
      />
    </WithStyles(ForwardRef(FormControl))>
    <WithStyles(ForwardRef(FormControl))
      className="Profile__FormControl"
      fullWidth={true}
    >
      <WithStyles(ForwardRef(InputLabel))
        htmlFor="maxAdverts"
      >
        Anuncios por p√°gina (Home)
      </WithStyles(ForwardRef(InputLabel))>
      <WithStyles(ForwardRef(Input))
        max={20}
        min={1}
        name="maxAdverts"
        onChange={[Function]}
        type="number"
        value={4}
      />
    </WithStyles(ForwardRef(FormControl))>
    <div
      className="Profile__Footer"
    >
      <WithStyles(ForwardRef(Button))
        className="ButtonWallakeep ButtonWallakeep__Green"
        color="primary"
        startIcon={<Memo(SaveIcon) />}
        type="submit"
        variant="contained"
      >
        Guardar
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        color="secondary"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        startIcon={<Memo(DeleteIcon) />}
        type="button"
        variant="contained"
      >
        Borrar
      </WithStyles(ForwardRef(Button))>
    </div>
  </form>
</Layout>
`;
